### PR TITLE
[flutter_tools] Make flutter git upstream configurable

### PR DESCRIPTION
**Pull request by TooYoungTooSimp**
## Description

Make flutter git upstream configurable from environment variable.

Used FLUTTER_GIT_URL to control _flutterGit value.

## Related Issues

#61286 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by TooYoungTooSimp** [comment]
@googlebot I signed it!

**Comment by TooYoungTooSimp** [comment]
Sorry for making many wrong commits at beginning. Please use rebase merge to hide them. 😢

**Comment by christopherfujino** [comment]
> Sorry for making many wrong commits at beginning. Please use rebase merge to hide them.

No prob, we default to squash and merge, so each PR becomes one commit on master :)

File: packages/flutter_tools/lib/src/version.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/version.dart:27]: This looks like a good solution. One thing though, is that if the user has a typo in this env variable (or if it the remote is down), we should explicitly show the override value in any error message. If you check line 654 of this file, we throw an error if a command we run does not return a 0 exit code. Even though this error includes the complete command that failed, unfortunately, if `FLUTTER_GIT_URL` env variable is wrong, we won't hit an error until line 257, which tries to run the command `git fetch $_versionCheckRemote $branch`, which doesn't include the upstream URL in it. Probably the better version would be to add a `catch` around 262, and IF we have an explicit `FLUTTER_GIT_URL` override set in the environment, `globals.printError('Warning: the Flutter git upstream was overriden by the environment variable FLUTTER_GIT_URL = ${globals.platform.environment['FLUTTER_GIT_URL']}');` (something like that, I didn't test it).

Also, this change should have unit tests in the file: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/version_test.dart. Using the `testUsingContext()` function, you can provide a `FakePlatform` and `ProcessManager` overrides (if you search the codebase for `FakePlatform`, you'll find a lot of examples) where you set the environment variable and then expect that you are getting the correct git commands to the `mockProcessManager`, specifically the one setting the upstream.

Let me know if you you run into any obstacles!

Comment (jmagman) [inline packages/flutter_tools/lib/src/version.dart:27]: Unfortunately this isn't really testable since it can't be changed after the first time it's set.  I wrote a unit test for this that passed on its own, but failed when run with the entire test suite because it always returned `https://github.com/flutter/flutter.git` because a previous test had initialized it.

How about:
```dart
String get _flutterGit => globals.platform.environment['FLUTTER_GIT_URL'] ??
    'https://github.com/flutter/flutter.git';
```

**Review by jmagman — CHANGES_REQUESTED** [review]
Thanks for your contribution!  This needs a test in order to be merged.  How about something like this in version_test.dart?

```dart
  testUsingContext('determine uses overridden git url', () {
    final MockProcessUtils processUtils = MockProcessUtils();
    when(processUtils.runSync(
      <String>['git', 'rev-parse', '--abbrev-ref', 'HEAD'],
      workingDirectory: anyNamed('workingDirectory'),
      environment: anyNamed('environment'),
    )).thenReturn(RunResult(ProcessResult(108, 0, 'master', ''), <String>['git', 'fetch']));
    when(processUtils.runSync(
      <String>['git', 'fetch', 'https://githubmirror.com/flutter.git', '--tags'],
      workingDirectory: anyNamed('workingDirectory'),
      environment: anyNamed('environment'),
    )).thenReturn(RunResult(ProcessResult(109, 0, '', ''), <String>['git', 'fetch']));
    when(processUtils.runSync(
      <String>['git', 'tag', '--contains', 'HEAD'],
      workingDirectory: anyNamed('workingDirectory'),
      environment: anyNamed('environment'),
    )).thenReturn(
        RunResult(ProcessResult(110, 0, '', ''),
          <String>['git', 'tag', '--contains', 'HEAD'],
        ));
    when(processUtils.runSync(
      <String>['git', 'describe', '--match', '*.*.*-*.*.pre', '--first-parent', '--long', '--tags'],
      workingDirectory: anyNamed('workingDirectory'),
      environment: anyNamed('environment'),
    )).thenReturn(RunResult(ProcessResult(111, 0, 'v0.1.2-3-1234abcd', ''), <String>['git', 'describe']));

    GitTagVersion.determine(processUtils, workingDirectory: '.', fetchTags: true);

    verify(processUtils.runSync(
      <String>['git', 'fetch', 'https://githubmirror.com/flutter.git', '--tags'],
      workingDirectory: anyNamed('workingDirectory'),
      environment: anyNamed('environment'),
    ));
  }, overrides: <Type, Generator>{
    Platform: () => FakePlatform(environment: <String, String>{
      'FLUTTER_GIT_URL': 'https://githubmirror.com/flutter.git',
    }),
  });
```
Those tests are kind of hard to write, sorry about that.  I'd be happy to push the test to your fork, if you'd like!

https://github.com/flutter/flutter/issues/61286#issuecomment-658565906
> Except flutter-build, all other test passed.
> To be honest, I don't know what is it and other prs have this problem too.

flutter-build is our post-submit CI system, and at the moment it's failing, but this has nothing to do with your PR.  Once this is in good shape, we'll handle merging your PR once it goes green!

File: packages/flutter_tools/lib/src/version.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/version.dart:27]: Good catch

**Comment by TooYoungTooSimp** [comment]
Really great. Thanks for all of your kindly helps. I'll work on them right now.
// Learned a lot this time 😃

**Comment by TooYoungTooSimp** [comment]
I got trouble writing tests...sorry.
After I copied the test from @jmagman and pushed to my branch.
Some CI tests failed.
Which result is
```
[test/general.shard/version_test.dart:main]:  NoSuchMethodError: The getter 'stdout' was called on null.
Receiver: null
Tried calling: stdout[test/general.shard/version_test.dart:main]:  #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
#1      _runGit (package:flutter_tools/src/version.dart:676:5)
#2      GitTagVersion.determine (package:flutter_tools/src/version.dart:760:9)
#3      main.<anonymous closure> (org-dartlang-app:///test/general.shard/version_test.dart:640:19)
#4      testUsingContext.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (org-dartlang-app:///test/src/context.dart:151:42)
#5      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:29)
```
And so on, but I believe these should be enough.
[detailed CI status](https://cirrus-ci.com/task/5008280592318464?command=main#L224)

So basically, I need your help.  😢

File: packages/flutter_tools/lib/src/version.dart
Comment (TooYoungTooSimp) [reply packages/flutter_tools/lib/src/version.dart]: I have finished print warning and unit test. But test running failed unexpected. Sorry for troubling but I have a problem now.

**Comment by TooYoungTooSimp** [comment]
Finally I managed to pass CI with tests.
So I would like to request you to review my pr again now.
@christopherfujino @jmagman

File: packages/flutter_tools/lib/src/version.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/version.dart:262]: nit: Missing a space between `catch` and `(`
```suggestion
    } on VersionCheckError catch (error) {
```

**Review by jmagman — COMMENTED** [review]
Other than the one spacing nit, this LGTM.  I'll approve and merge once that gets fixed, assuming @christopherfujino doesn't have any more suggestions.

**Comment by christopherfujino** [comment]
LGTM once you make @jmagman 's spacing change

**Comment by TooYoungTooSimp** [comment]
The space problem fixed. Thanks.

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**
